### PR TITLE
Use ConditionalResourceProvider in the Spring Boot service name provider

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**

### CODE REVIEW

**Review by laurit — APPROVED** [review]

File: instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java [hunk 1]
```diff
@@ -95,6 +96,21 @@ public Resource createResource(ConfigProperties config) {
         .orElseGet(Resource::empty);
   }
 
+  @Override
+  public boolean shouldApply(ConfigProperties config, Resource resource) {
+    String serviceName = config.getString("otel.service.name");
+    Map<String, String> resourceAttributes = config.getMap("otel.resource.attributes");
+    return serviceName == null
+        && !resourceAttributes.containsKey(ResourceAttributes.SERVICE_NAME.getKey())
+        && "unknown_service:java".equals(resource.getAttribute(ResourceAttributes.SERVICE_NAME));
```
Comment (trask) [inline instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java:105]: can you add a comment to explain that checking otel.service.name and otel.resource.attributes is just done for performance reasons (I think)?

**Review by trask — APPROVED** [review]

File: instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java (hunk 1 above)
Comment (breedx-splk) [reply instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java:105]: IIRC, It's not entirely for performance (although that's also true). This strategy is a "best guess" set of heuristics and we don't want to ever override a name applied by a previous strategy that might know better.

**Review by breedx-splk — APPROVED** [review]
Looks good to me. Thanks for following this up.

File: instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java (hunk 1 above)
Comment (trask) [reply instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java:105]: I believe `otel.service.name` and `otel.resource.attributes` are applied after all of the user resource providers, so those would always take precedence, and so I think no need to check them here then?

Comment (breedx-splk) [reply instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java:105]: Makes sense, then this does indeed help to guard against attempting to load resources from disk and classpath.

Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java:107]: Added comment 👍

**Pull request merged**